adding Vercel ai sdk llm docs (#1719)
SigNoz/signoz.io: data/docs/llm/vercel-ai-sdk-monitoring.mdx

---
date: 2025-07-21

> id: vercel-ai-sdk-monitoring

Reviewer: ```suggestion
id: vercel-ai-sdk-monitoring-with-opentelemetry
```

Reviewer: This helps to make the URL more SEO friendly if someone is searching for vercel ai ask monitoring with opentelemetry.

---

## Overview

This guide walks you through enabling telemetry in the Vercel AI SDK and streaming that data to SigNoz Cloud using OpenTelemetry. By the end of this setup, you'll be able to monitor AI-related function calls like generateText() or generateStream()—including request durations, inputs, outputs, and custom metadata—directly within SigNoz.

Instrumenting your AI workflows with telemetry helps you observe performance, debug latency, and trace AI interactions end-to-end across your Next.js application. This is especially valuable when building production-grade AI features where visibility into model behavior and user interactions is critical.

## Prerequisites

- Next.js app
- Vercel AI SDK integrated into the app
- [SigNoz Cloud Account](https://signoz.io/teams/)
- SigNoz Ingestion Key

## Instrument your Next.js application

Check out detailed instructions on how to set up OpenTelemetry instrumentation in your Nextjs applications and view your application traces in SigNoz over [here](https://signoz.io/docs/instrumentation/opentelemetry-nextjs/).

## Send traces directly to SigNoz Cloud

**Step 1.** Install OpenTelemetry packages

> ```bash
> npm install @vercel/otel @opentelemetry/api
> 

Reviewer: ```suggestion
```bash
npm install @vercel/otel @opentelemetry/api
```

    }),
  });
}

```

- **`<service_name>`** is the name of your service
- Set the **`<region>`** to match your SigNoz Cloud [**region**](https://signoz.io/docs/ingestion/signoz-cloud/overview/#endpoint)
- Replace **`<your-ingestion-key>`** with your SigNoz [**ingestion key**](https://signoz.io/docs/ingestion/signoz-cloud/keys/)

> The instrumentation file should be in the root of your project and not inside the app or pages directory. If you're using the src folder, then place the file inside src alongside pages and app.
> 

Your Next.js app should be properly instrumented now.

**Step 4.** Verify Instrumentation Locally

Run your Next.js application in development mode:

```bash
npm run dev
```

In your terminal, you should see output from OpenTelemetry when the application starts, confirming that the instrumentation.ts file was loaded correctly. 

![Terminal output showing OpenTelemetry initialization](/img/docs/llm/vercel/vercel-run-output.webp)

To see detailed trace export logs, you can temporarily set the DiagLogLevel to DEBUG in instrumentation.ts:

```jsx
diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);
```

## Enable Telemetry for Vercel AI SDK

> The Vercel AI SDK uses [OpenTelemetry](https://opentelemetry.io/) to collect telemetry data. OpenTelemetry is an open-source observability framework designed to provide standardized instrumentation for collecting telemetry data.

Reviewer: ```suggestion
The Vercel AI SDK uses [OpenTelemetry](https://signoz.io/blog/what-is-opentelemetry/) to collect telemetry data. OpenTelemetry is an open-source observability framework designed to provide standardized instrumentation for collecting telemetry data.
```

## Telemetry Metadata

You can provide a `functionId` to identify the function that the telemetry data is for, and `metadata` to include additional information in the telemetry data.

```jsx
const result = await generateText({
  model: openai('gpt-4-turbo'),
  prompt: 'Write a short story about a cat.',
  experimental_telemetry: {
    isEnabled: true,
    functionId: 'my-awesome-function',
    metadata: {
      something: 'custom',
      someOtherThing: 'other-value',
    },
  },
});
```

## Custom Tracer

You may provide a `tracer` which must return an OpenTelemetry `Tracer`. This is useful in situations where you want your traces to use a `TracerProvider` other than the one provided by the `@opentelemetry/api` singleton.

```jsx
const tracerProvider = new NodeTracerProvider();
const result = await generateText({
  model: openai('gpt-4-turbo'),
  prompt: 'Write a short story about a cat.',
  experimental_telemetry: {
    isEnabled: true,
    tracer: tracerProvider.getTracer('ai'),
  },
});
```

> Your Vercel AI SDK commands should now automatically emit traces, spans, and events according to [here](https://ai-sdk.dev/docs/ai-sdk-core/telemetry#collected-data). 

Reviewer: ```suggestion
Your Vercel AI SDK commands should now automatically emit traces, spans, and events. You can find more details on the types of spans and events generated [here](https://ai-sdk.dev/docs/ai-sdk-core/telemetry#collected-data).
```

Finally, you should be able to view this data in Signoz Cloud under the traces tab:

> ![Traces View](/img/docs/llm/vercel/vercel-traces-view.webp)

Reviewer: <Figure src="/img/docs/llm/vercel/vercel-traces-view.webp" alt="See spans in SigNoz" caption="See collected spans in SigNoz" />

@gkarthi-signoz let's use this component for image - and set alt text and caption for images.

Reviewer: <Figure src="/img/docs/llm/vercel/vercel-traces-view.webp" alt="Collected spans in SigNoz" caption="Collected spans in SigNoz" />

> ![Detailed Traces View](/img/docs/llm/vercel/vercel-traces-detailed-view.webp)

Reviewer: ```suggestion
If you click on any span, you can see the detailed trace of which it is a part of.

![Detailed Traces View](/img/docs/llm/vercel/vercel-traces-detailed-view.webp)
```
